framework/src/org/checkerframework/common/signatureinference/SignatureInferenceScenes.java

```diff
@@ -0,0 +1,588 @@
+     * type hierarchy will be used to update the method's return type.
+     */
+    public static void updateMethodReturnTypeInScene(ReturnNode retNode,
+            ClassSymbol classSymbol, MethodTree methodTree,
+            AnnotatedTypeFactory atf) {
+        if (classSymbol == null) return;
+        AnnotatedTypeMirror methodReturnType = atf.getAnnotatedType(methodTree).
+                getReturnType();
+        if (methodReturnType.getExplicitAnnotations().size() > 0) {
+            // We do not infer types if there are explicit annotations.
+            // See https://github.com/typetools/annotation-tools/issues/105
+            return;
+        }
+        String jaifPath = JAIF_FILES_PATH + classSymbol.flatname.toString() +
+                ".jaif";
+        AScene scene = getScene(jaifPath);
+        AnnotatedTypeMirror returnExprATM = atf.getAnnotatedType(retNode.
+                getTree().getExpression());
+
+        AClass clazz = getJaifClass(classSymbol, scene);
+        if (clazz == null) return; // Anonymous class => Ignore, for now.
+        String methodName = JVMNames.getJVMMethodName(methodTree);
+        AMethod method = getJaifMethod(clazz, methodName, scene);
+        if (method.returnType != null) {
+            updateAnnotationSetInScene(
+                    method.returnType.tlAnnotationsHere, atf,
+                    jaifPath, returnExprATM);
+        }
+    }
+
+    /**
+     * Updates the set of annotations in a location of a Scene. If the set of
+     * annotations (curAnnos) is empty, then the updated set will be the set of
+     * annotations in newATM.
+     * If curAnnos is not empty, the updated set will be the LUB between
+     * curAnnos and newATM.
+     *
+     * @param curAnnos set of annotations located somewhere in a Scene.
+     * @param atf the annotated type factory of a given type system, whose
+     * type hierarchy will be used to update curAnnos.
+     * @param jaifPath identifies a Scene.
+     * @param newATM the type containing a set of annotations to be
+     * added (or "lubbed") to the Scene.
+     */
+    private static void updateAnnotationSetInScene(Set<Annotation> curAnnos,
+            AnnotatedTypeFactory atf, String jaifPath,
+            AnnotatedTypeMirror newATM) {
+        if (curAnnos != null && curAnnos.size() > 0) {
+            AnnotatedTypeMirror curAnnosATM = setOfAnnotationsToATM(
+                    curAnnos, atf, false, newATM.getUnderlyingType());
+            if (curAnnosATM != null) {
+                newATM = AnnotatedTypes.leastUpperBound(
+                        atf.getProcessingEnv(), atf, newATM, curAnnosATM);
+            }
+        }
+
+        Set<Annotation> setOfAnnos = atmToSetOfAnnotations(newATM, atf);
+        curAnnos.clear();
+        curAnnos.addAll(setOfAnnos);
+        addModifiedScene(jaifPath);
+    }
+
+    /**
+     * Returns the AField in an AScene, given an AClass and a FieldAccessNode.
```

Comment:
This method and the ones below are different than some other "get" methods.  "getScene" reads a file if necessary.  However, this method does not -- it makes the assumption that the file has already been read.
Please document this assumption.
Actually, would it be simpler for this class to simply read all the .jaif files first, before doing any other actions, rather than reading each one on-demand?  On-demand might be more efficient (if in many cases only a few are ever read), but it does complicate the logic.  Can you expand on the design choices?